activerecord/lib/active_record/filter_attribute_handler.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class FilterAttributeHandler # :nodoc:
+    class << self
+      def on_sensitive_attribute_declared(&block)
+        @encrypted_attribute_declaration_listeners ||= Concurrent::Array.new
+        @encrypted_attribute_declaration_listeners << block
+      end
+
+      def sensitive_attribute_was_declared(klass, list)
+        @encrypted_attribute_declaration_listeners&.each do |block|
+          block.call(klass, list)
+        end
+      end
+    end
+
+    def initialize(app)
+      @app = app
+      @attributes_by_class = Concurrent::Map.new
+      @collecting = true
+
+      install_collecting_hook
+    end
+
+    def enable
+      apply_collected_attributes
+      @collecting = false
+    end
+
+    private
+      attr_reader :app
+
+      def install_collecting_hook
+        self.class.on_sensitive_attribute_declared do |klass, list|
+          attribute_was_declared(klass, list)
+        end
+      end
+
+      def attribute_was_declared(klass, list)
+        if collecting?
+          collect_for_later(klass, list)
+        else
+          apply_filter(klass, list)
+        end
+      end
+
+      def apply_collected_attributes
+        @attributes_by_class.each do |klass, list|
+          apply_filter(klass, list)
+        end
+      end
+
+      def collecting?
+        @collecting
+      end
+
+      def collect_for_later(klass, list)
+        @attributes_by_class[klass] ||= Concurrent::Array.new
+        @attributes_by_class[klass] += list
+      end
+
+      def apply_filter(klass, list)
+        list.each do |attribute|
+          klass_name = (klass.name && klass < ActiveRecord::Base) ? klass.model_name.element : nil
+          filter = [klass_name, attribute.to_s].compact.join(".")
+          app.config.filter_parameters << filter unless app.config.filter_parameters.include?(filter)
```

Comment:
Do you think it's worth re-compiling the filters?

The filters are precompiled and memoized on the first request, models that get loaded after the memoization happen (this is mostly in development when things are not eagerloaded) will not be precompiled. This may result in having an increasing number of filters to iterate through each time a new model is loaded.

https://github.com/rails/rails/blob/3954e4681de1013d1dc2efc166fb765be62650bd/railties/lib/rails/application.rb#L320-L321
https://github.com/rails/rails/blob/3954e4681de1013d1dc2efc166fb765be62650bd/railties/lib/rails/application.rb#L654-L659